### PR TITLE
feat: Add ability to define pixel difference calculators

### LINKED ISSUE #16: Transparent pixels with different RGB values fail on assertSimilarGD
An issue I've seen crop up from time to time when using this library is when transparent pixels between two images (ie. alpha = 127)  have differing RGB values. When viewing the images, they are exactly the same, for example:

transparent-black
transparent-white

Both two images look exactly the same, but the top one has a transparent black background around the edge (RGB 0) and the bottom one has a transparent white background around the edge (RGB 255). Due to the RGB values being different, all these edge pixels are considered different, and thus the threshold of the test has to set to a threshold high enough to cover the percentage of the image that these pixels take up.

I have experimented applying a different colour difference calculation that takes the alpha channel into account when calculating the difference of the RGB channels and basically use a percentage of the RGB values depending on the opacity - for example, at 50% opacity, the RGB values are halved and thus the delta of difference is less, and at 0% opacity, the RGB values are considered to be 0 themselves, meaning not taken into account at all for difference. This started to go into the realm of determining differences based on opacity which might be outside of the scope for this library.

I'm interested to hear your thoughts on whether you would be amenable to a PR that includes such calculations, or if for now, a smaller scoped PR that just ignores RGB values when the pixel is transparent?

### PR DESCRIPTION
**Pull request by bennothommo**
As discussed in https://github.com/meyfa/phpunit-assert-gd/issues/16, this adds the ability to define calcuators for determining the difference between two images at given pixel co-ordinates. It maintains backwards compatibility by using the current method of pixel difference calculation and adding an additional optional parameter to the `assertSimilarGD` and `assertNotSimilarGD` methods.

Users of this library will be able to define custom calculation methods by implementing the `AssertGD\DiffCalculator` interface and providing an instance of the custom calculator as the 5th parameter to the above methods.

It should hopefully be written in a way that it can be backported to older versions of this library. Happy to change the naming if there are better alternatives.

### CODE REVIEW

**Comment by meyfa** [comment]
Thanks so much, this looks good to me! Nicely done.

I think this needs some documentation though, would you like to add that to the README?

**Comment by bennothommo** [comment]
Of course, I'll try and sort that out in the next day or two.

**Comment by bennothommo** [comment]
Sorry for the delay @meyfa - documentation has now been added to the README.

File: README.md [hunk 1]
```diff
@@ -89,3 +89,52 @@ $this->assertThat(
...
 ```
+
+## Difference calculation
+
+By default, this library calculates the difference between two images by
+comparing the RGBA color channel information at each pixel coordinate of the
+source image and the test image, and averaging the difference between each
+pixel to calculate the difference score.
+
+This will work for the majority of cases, but may given incorrect scoring
```
Comment (meyfa) [inline README.md]:
```suggestion
This will work for the majority of cases, but may give incorrect scoring 
```

**Comment by meyfa** [comment]
@bennothommo No worries!

This is really awesome, thank you so much for this great PR!

**Comment by meyfa** [comment]
Released as v4.1.0 🎉 - and backported to v3.1.0, v2.2.0, v1.3.0.

**Pull request merged**